base_layer/core/src/validation/aggregate_body/aggregate_body_internal_validator.rs

```diff
@@ -0,0 +1,522 @@
+        kernel.verify_signature().map_err(|e| {
+            warn!(target: LOG_TARGET, "Kernel ({}) signature failed {:?}.", kernel, e);
+            e
+        })?;
+    }
+    Ok(())
+}
+
+/// Confirm that the (sum of the outputs) - (sum of inputs) = Kernel excess
+///
+/// The offset_and_reward commitment includes the offset & the total coinbase reward (block reward + fees for
+/// block balances, or zero for transaction balances)
+fn validate_kernel_sum(
+    body: &AggregateBody,
+    offset_and_reward: Commitment,
+    factory: &CommitmentFactory,
+) -> Result<(), ValidationError> {
+    trace!(target: LOG_TARGET, "Checking kernel total");
+    let KernelSum { sum: excess, fees } = sum_kernels(body, offset_and_reward);
+    let sum_io = sum_commitments(body)?;
+    trace!(target: LOG_TARGET, "Total outputs - inputs:{}", sum_io.to_hex());
+    let fees = factory.commit_value(&PrivateKey::default(), fees.into());
+    trace!(
+        target: LOG_TARGET,
+        "Comparing sum.  excess:{} == sum {} + fees {}",
+        excess.to_hex(),
+        sum_io.to_hex(),
+        fees.to_hex()
+    );
+    if excess != &sum_io + &fees {
+        return Err(ValidationError::CustomError(
+            "Sum of inputs and outputs did not equal sum of kernels with fees".into(),
+        ));
+    }
+
+    Ok(())
+}
+/// Calculate the sum of the kernels, taking into account the provided offset, and their constituent fees
+fn sum_kernels(body: &AggregateBody, offset_with_fee: PedersenCommitment) -> KernelSum {
+    // Sum all kernel excesses and fees
+    body.kernels().iter().fold(
+        KernelSum {
+            fees: MicroTari(0),
+            sum: offset_with_fee,
+        },
+        |acc, val| KernelSum {
+            fees: acc.fees + val.fee,
+            sum: &acc.sum + &val.excess,
+        },
+    )
+}
+
+/// Calculate the sum of the outputs - inputs
+fn sum_commitments(body: &AggregateBody) -> Result<Commitment, ValidationError> {
+    let sum_inputs = body
+        .inputs()
+        .iter()
+        .map(|i| i.commitment())
+        .collect::<Result<Vec<&Commitment>, _>>()?
+        .into_iter()
+        .sum::<Commitment>();
+    let sum_outputs = body.outputs().iter().map(|o| &o.commitment).sum::<Commitment>();
+    Ok(&sum_outputs - &sum_inputs)
+}
+
+fn validate_range_proofs(
+    body: &AggregateBody,
+    range_proof_service: &RangeProofService,
+) -> Result<(), TransactionError> {
+    trace!(target: LOG_TARGET, "Checking range proofs");
+    let outputs = body.outputs().iter().collect::<Vec<_>>();
+    batch_verify_range_proofs(range_proof_service, &outputs)?;
+    Ok(())
+}
+
+fn verify_metadata_signatures(body: &AggregateBody) -> Result<(), ValidationError> {
+    trace!(target: LOG_TARGET, "Checking sender signatures");
+    for o in body.outputs() {
+        o.verify_metadata_signature()?;
+    }
+    Ok(())
+}
+
+/// this will validate the script offset of the aggregate body.
+fn validate_script_offset(
```

Comment:
```suggestion
/// this will validate the script and script offset of the aggregate body.
fn validate_script_and_script_offset(
```
nit, this does both script validation and script offset validation. 
Script validation can be done without script offset validation, but script offset cannot be done without running the script. So validating the offset, means the script must also be checked, but this is detailed knowledge of TariScript, that should not be required for reading the code.